### PR TITLE
State Transfer Improvements

### PR DESCRIPTION
**Pull request by glevkovich**
This branch aims to improve State's Transfer (ST) performance and save significant disk space during the State-Transfer cycle. Until now,(ST would not be able to perform if the amount of total disk space require to fetch is greater than approx 50% of the left disk space).  Changes in short:

- Three main stages until a block is part of the consensus blockchain: Fetch, Commit, Post-Process.
- ST Post-processing is decoupled from ST networking fetch and commit (separate threads and logic).
- Blocks are collected in small chunks (Fetch Ranges) from oldest to newest.
- The above changes allow the post-processing thread to start at the same time as the networking thread, so we get a performance improvement in the total ST cycle time.
- To apply BFT security - we are using an RVT (Range Validation Tree) which helps to validate all the blocks in a fetching range by validating a single block (RVB block) sent from a single source. The tree is validated from f+1 replicas. This prevents a byzantine source(s) from performing a DOS attack on ST: sending incorrect blocks may be possible only for a short time, ST will be able to track quickly such source and replace it with "an honest" source.

Some comments:
- There is a known issue in the storage layer - kv_blockchain* and direct_kv_db_adapter*. Please do not review these parts yet.
- Few separate commits are meant to stay only for the dev branch - commenting apollo tests, modifying the Github actions workflow - they will be dropped before the merge.
- This merge message will be modified before the final merge.

### CODE REVIEW

File: bftengine/src/bcstatetransfer/Messages.hpp [hunk 1]
```diff
@@ -53,56 +55,110 @@ struct AskForCheckpointSummariesMsg : public BCStateTranBaseMsg {
...
 
 struct CheckpointSummaryMsg : public BCStateTranBaseMsg {
-  CheckpointSummaryMsg() {
-    // NOLINTNEXTLINE(bugprone-undefined-memory-manipulation)
-    memset(this, 0, sizeof(CheckpointSummaryMsg));
-    type = MsgType::CheckpointsSummary;
+  CheckpointSummaryMsg() = delete;
+
+  static CheckpointSummaryMsg* create(size_t rvbDataSize) {
+    size_t totalSize = sizeof(CheckpointSummaryMsg) + rvbDataSize - 1;
```
Comment (WildFireFlum) [inline bftengine/src/bcstatetransfer/Messages.hpp:61]: Maybe rename to totalByteSize (and the parameter as well, as it is not clear what size unit it uses)?
What is the minus 1 for?
Why not change the data member to a unique ptr of a byte array?

File: bftengine/src/bcstatetransfer/DBDataStore.cpp [hunk 2]
```diff
@@ -25,6 +26,37 @@ std::string toString(const DataStore::CheckpointDesc& desc) {
 }
 
 /** ******************************************************************************************************************/
+void DBDataStore::memoryStateToLog() {
+  std::ostringstream oss;
+
+  const auto replicas = inmem_->getReplicas();
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/DBDataStore.cpp:32]: I guess having that is better, but this code partially/mostly just moved and being refactored , and is using existing helpers.
I do not have the time to improve what we already have.

File: bftengine/src/bcstatetransfer/DBDataStore.cpp [hunk 3]
```diff
@@ -147,15 +176,17 @@ void DBDataStore::setReplicas(const std::set<std::uint16_t> replicas) {
  */
 void DBDataStore::serializeCheckpoint(std::ostream& os, const CheckpointDesc& desc) const {
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/DBDataStore.cpp:177]: As far as I understand there are 2 options to do things:

1. If we expect a type to be used in few/many modules we will write a Serializable::serialize under the matching serialization namespace. In some cases we might even write it in the dedicated file.
2. In this case the type is used mostly by ST, and is not a native/widely used type. We are using the existing functions to serialize CheckpointDesc.

Furthermore, this is an already existing code perior to the PR.

File: bftengine/src/bcstatetransfer/DBDataStore.cpp [hunk 4]
```diff
@@ -147,15 +176,17 @@ void DBDataStore::setReplicas(const std::set<std::uint16_t> replicas) {
...
   Serializable::serialize(os, desc.checkpointNum);
-  Serializable::serialize(os, desc.lastBlock);
-  Serializable::serialize(os, desc.digestOfLastBlock.get(), BLOCK_DIGEST_SIZE);
+  Serializable::serialize(os, desc.maxBlockId);
+  Serializable::serialize(os, desc.digestOfMaxBlockId.get(), BLOCK_DIGEST_SIZE);
   Serializable::serialize(os, desc.digestOfResPagesDescriptor.get(), BLOCK_DIGEST_SIZE);
+  Serializable::serialize(os, desc.rvbData);
 }
 void DBDataStore::deserializeCheckpoint(std::istream& is, CheckpointDesc& desc) const {
   Serializable::deserialize(is, desc.checkpointNum);
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/DBDataStore.cpp:185]: same as above.

File: bftengine/src/bcstatetransfer/Messages.hpp (hunk 1 above)
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/Messages.hpp:61]: - totalByteSize -> OK thanks, I'll change it.
- What is the minus 1 for -> substruct data[1] (dynamic array size)
- Having a vector gives many multiple advantages, for example I do not need to check the size, is it empty and many other things )I get many simple API functions which simplify the code) In general, vector is a one of the best solutions after a unique pointer to keep a buffer - while it also manages its own memory.

File: bftengine/src/bcstatetransfer/DBDataStore.cpp [hunk 5]
```diff
@@ -432,6 +464,39 @@ void DBDataStore::clearDataStoreData() {
   inmem_.reset(new InMemoryDataStore(inmem_->getSizeOfReservedPage()));
 }
 
+/** ******************************************************************************************************************
+ *  Range Validation Blocks
+ */
+
+void DBDataStore::serializePrunedBlocksDigests(std::ostream& os,
+                                               const std::vector<std::pair<BlockId, STDigest>>& digests) const {
+  Serializable::serialize(os, digests.size());
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/bcstatetransfer/DBDataStore.cpp:473]: I was aware of that, and due to time restriction decided to use the building blocks I already have.
With the time frame we have, the above solution is reasonble, peofromance improvment is neglactable.
Its a complex type, I prefer to serialize it with the small building blocks I have.
This is already done extensively anywhere throughout the code.

**Review by mdarade — APPROVED** [review]
Looks good.
Thanks Gil.

**Pull request merged**